distributed/protocol/computation.py

```diff
@@ -0,0 +1,260 @@
+"""
+This module implements graph computations based on the specification in Dask[1]:
+> A computation may be one of the following:
+>   - Any key present in the Dask graph like `'x'`
+>   - Any other value like `1`, to be interpreted literally
+>   - A task like `(inc, 'x')`
+>   - A list of computations, like `[1, 'x', (inc, 'x')]`
+
+In order to support efficient and flexible task serialization, this module introduces
+classes for computations, tasks, data, functions, etc.
+
+Notable Classes
+---------------
+
+- `PickledObject` - An object that are serialized using `protocol.pickle`.
+  This object isn't a computation by itself instead users can build pickled
+  computations that contains pickled objects. This object is automatically
+  de-serialized by the Worker before execution.
```

Comment:
Should we have PickledObjects?  Or should we use the general serialization path for this?

Reply:
You mean if we have like a list of objects? That should also just use a single `PickledObject` to serialize everything in one go.
We use `typeset_computation()` to look through a computation and wrap individual task functions in `PickledCallable`.

Reply:
I mean "why would we ever want to pickle an object, rather than give the rest of our serialization machinery a chance to work?"  For example, what if the object was a cupy array.  

Or maybe this is used very infrequently, and not where user-data is likely to occur?

Reply:
This is basically the fine vs coarse grained serialization discussion. This PR continues the existing coarse grained approach where (nested) tasks are pickled. We used to use `dumps_function()` and `warn_dumps()` to do this. 
Now, we use `PickledComputation` as the _outermost_ wrapper and `PickledObjects` for already pickled objects. This makes it possible for HLG.unpack() that runs on the Scheduler to build new tasks of already pickled objects.

Reply:
> Or maybe this is used very infrequently, and not where user-data is likely to occur?

Yes, including large data in the task directly will raise a warning just like it use to.